### PR TITLE
[ROCm] Add a first version of bf16 atomicAdd

### PR DESCRIPTION
**Pull request by iotamudelta**
This ROCm-only change will eventually be replaced with a HIP-level API. In the meantime, do our own CAS loop inside PT.

### CODE REVIEW

File: aten/src/THC/THCAtomics.cuh [hunk 1]
```diff
@@ -103,6 +103,24 @@ static inline __device__ void atomicAdd(bool *address, bool val) {
   *address = address && val;
 }
 
+#ifdef __HIP_PLATFORM_HCC__
```
Comment (bddppq) [inline aten/src/THC/THCAtomics.cuh]: I don't see any ROCm specific api in the code, so I think it should work on cuda as well?

**Review by bddppq — COMMENTED** [review]
Could you add/enable some tests?

File: aten/src/THC/THCAtomics.cuh (hunk 1 above)
Comment (iotamudelta) [reply aten/src/THC/THCAtomics.cuh]: AFAIK there is no bf16 support for CUDA yet?

**Comment by iotamudelta** [comment]
Tests will require more ops to be enabled for bf16 .

File: aten/src/THC/THCAtomics.cuh (hunk 1 above)
Comment (bddppq) [reply aten/src/THC/THCAtomics.cuh]: `at::BFloat16` should work on CUDA

Can you try removing the ifdef and see if cuda CI jobs complain?

**Review by izdeby — CHANGES_REQUESTED** [review]
> Tests will require more ops to be enabled for bf16 .

In this case we should consider either enable those ops as a part of this PR or make a stack. We shouldn't add code which is not covered at least with some basic tests.

**Comment by iotamudelta** [comment]
CUDA doesn't work.

File: aten/src/THC/THCAtomics.cuh (hunk 1 above)
Comment (izdeby) [reply aten/src/THC/THCAtomics.cuh]: There is basic support. Please see [PR](https://github.com/pytorch/pytorch/pull/27259)

**Comment by iotamudelta** [comment]
Superseeded by #34166

**Pull request closed without merging**